File: 0289-game-of-life/0289-game-of-life.go
func gameOfLife(board [][]int) {
    rows := len(board)
    cols := len(board[0])
    for i := 0; i < rows; i++ {
        for j := 0; j < cols; j++ {
            neighbours := countNeighbours(board, i, j)
            if board[i][j] == 1 {
                if neighbours < 2 || neighbours > 3 {
                    board[i][j] = 2
                }
            } else {
                if neighbours == 3 {
                    board[i][j] = -1
                }
            }
        }
    }
    for i := 0; i < rows; i++ {
        for j := 0; j < cols; j++ {
            if board[i][j] == 2 {
                board[i][j] = 0
            } else if board[i][j] == -1 {
                board[i][j] = 1
            }
        }
    }
}
func countNeighbours(board [][]int, x, y int) int {
    count := 0
    rows := len(board)
    cols := len(board[0])
    for _, dir := range [][]int{
        {-1, -1}, {-1, 0}, {-1, 1},
        {0, -1}, {0, 1},
        {1, -1}, {1, 0}, {1, 1},
    } {
        newX := x + dir[0]
        newY := y + dir[1]
        if newX >= 0 && newX < rows && newY >= 0 && newY < cols {
            if board[newX][newY] == 1 || board[newX][newY] == 2 {
                count++
            }
        }
    }
    return count
}

File: 0289-game-of-life/0289-game-of-life.java
class Solution {
    public void gameOfLife(int[][] board) {
        int[][] nextGen = new int[board.length][board[0].length];         
        for(int i = 0; i < board.length; i++){
            for(int j = 0; j < board[0].length; j++){
                int neighborCount = countNeighbors(board, i, j);                
                // Dead
                if(board[i][j] == 0){
                    if(neighborCount == 3){
                        nextGen[i][j] = 1;
                    }
                }else{
                    if(neighborCount < 2){
                        nextGen[i][j] = 0;
                    }else if(neighborCount < 4){
                        nextGen[i][j] = 1;
                    }else{
                        nextGen[i][j] = 0;
                    }
                }
            }
        }
        for(int i = 0; i < board.length; i++){
            for(int j = 0; j < board[0].length; j++){
                board[i][j] = nextGen[i][j];
            }
        }
    }
    private int countNeighbors(int[][] board, int y, int x){
        int count = 0;
        for(int i = -1; i <= 1; i++){
            for(int j = -1; j <= 1; j++){
                if(i == 0 && j == 0) continue;
                int newY = y + i;
                int newX = x + j; 
                if(isInBounds(board, newY, newX) && board[newY][newX] == 1){
                    count++;
                }
            }
        }
        return count;
    }

    private boolean isInBounds(int[][]board, int y, int x){
        if(x < 0 || y < 0 || x >= board[0].length || y >= board.length){
            return false;
        }
        return true;
    }
}